src/provider/mod.rs

```diff
@@ -31,6 +34,268 @@ pub struct Eip712Fee {
+    pub value: B256,
+    pub index: u64,
+}
+
+/// Response type for `zks_getProof`.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Proof {
+    pub address: Address,
+    pub storage_proof: Vec<StorageProof>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct StorageLog {
+    pub address: Address,
+    pub key: U256,
+    pub written_value: U256,
+}
+
+/// A log produced by a transaction.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Log {
+    pub address: Address,
+    pub topics: Vec<B256>,
+    pub data: Bytes,
+    pub block_hash: Option<B256>,
+    pub block_number: Option<U64>,
+    pub l1_batch_number: Option<U64>,
+    pub transaction_hash: Option<B256>,
+    pub transaction_index: Option<U64>,
+    pub log_index: Option<U64>,
+    pub transaction_log_index: Option<U64>,
+    pub log_type: Option<String>,
+    pub removed: Option<bool>,
+    pub block_timestamp: Option<U64>,
+}
+
+/// Response type for `zks_sendRawTransactionWithDetailedOutput`.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TransactionDetailedResult {
+    pub transaction_hash: B256,
+    pub storage_logs: Vec<StorageLog>,
+    pub events: Vec<Log>,
+}
+
+/// Response type for `zks_getConfirmedTokens`.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct Token {
+    pub l1_address: Address,
+    pub l2_address: Address,
+    pub name: String,
+    pub symbol: String,
+    pub decimals: u8,
+}
+
+/// Response type for `zks_getL2ToL1LogProof`.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct L2ToL1LogProof {
+    /// The merkle path for the leaf.
+    pub proof: Vec<B256>,
+    /// The id of the leaf in a tree.
+    pub id: u32,
+    /// The root of the tree.
+    pub root: B256,
+}
+
+type GetMsgProofRequest = (u64, Address, B256, Option<usize>);
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct Execute {
+    pub contract_address: Option<Address>,
+    pub calldata: Bytes,
+    pub value: U256,
+    /// Factory dependencies: list of contract bytecodes associated with the deploy transaction.
+    pub factory_deps: Vec<Bytes>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct InputData {
+    pub hash: B256,
+    pub data: Bytes,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(rename_all = "camelCase")]
+pub struct L2TxCommonData {
+    pub nonce: u32,
+    pub fee: Eip712Fee,
+    pub initiator_address: Address,
+    pub signature: Bytes,
+    pub transaction_type: String,
+    pub input: Option<InputData>,
+    pub paymaster_params: PaymasterParams,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum ExecuteTransactionCommon {
+    L2(L2TxCommonData),
+}
```

Comment:
I believe L1 and protocol upgrade transactions can also be queried through API.

Reply:
Good catch, I've added both.